### PR TITLE
ReplacementLayer.CantHappen

### LINKED ISSUE #2197: "Can't enter" effects (Moved Prevent$) mess up ChangedState
Examples with _Grafdigger's Cage_ in play:
- Cast _Ever After_ on _Abbey Griffin_. You now have a Zombie Griffin in your graveyard
- Try to return artifact creature that got Unearth from _Ghost Ark_. It now lost Unearth
- _Startled Awake_ will transform and then stay in the graveyard, but it should be allowed to ETB

Also combined with _Containment Priest_ Forge allows you to order both so you could make it exile which seems wrong too.

They should probably be done as statics?

### PR DESCRIPTION
**Pull request by tool4ever**
Closes #2197

### CODE REVIEW

File: forge-game/src/main/java/forge/game/GameAction.java [hunk 1]
```diff
@@ -365,18 +365,31 @@ private Card changeZone(final Zone zoneFrom, Zone zoneTo, final Card c, Integer
                 copied.getOwner().removeInboundToken(copied);
 
                 if (repres == ReplacementResult.Prevented) {
-                    if (game.getStack().isResolving(c) && !zoneTo.is(ZoneType.Graveyard)) {
+                    c.clearEtbCounters();
+                    if (cause != null) {
+                        if (cause.hasParam("AnimateSubAbility")) {
+                            c.removeChangedCardKeywords(game.getTimestamp(), 0);
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/GameAction.java:371]: i don't know if that is going to have the right timestamps

**Comment by Hanmac** [comment]
while touching ChangeEffect can you look at #1807 ? just remove Converted

For #1998 it Probably need to be moved to CardFactoryUtil so it is a layer before MayPlay stuff

**Comment by tool4ever** [comment]
> while touching ChangeEffect can you look at #1807 ? just remove Converted
> 
> For #1998 it Probably need to be moved to CardFactoryUtil so it is a layer before MayPlay stuff

Icky stuff, don't think I could do it partially without updating all cards in this

But I'll see if I can take care of it a little bit later

File: forge-game/src/main/java/forge/game/GameAction.java (hunk 1 above)
Comment (tool4ever) [reply forge-game/src/main/java/forge/game/GameAction.java:371]: You're right, some corner case might change it between
See what you think of this approach

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/GameAction.java]: imo that feels more save for now

**Review by kevlahnota — APPROVED** [review]

**Pull request merged**